specs/waku/v2/waku-store-pagination.md

```diff
@@ -0,0 +1,102 @@
+- [Wire Specification](#wire-specification)
+  * [Protobuf](#protobuf)
+
+# Abstract
+
+Pagination is an optional service enabled on top of the `WakuStore` protocol. By using pagination, a requester can set the number of waku messages to return and the offset to start from. The queried node MUST then use this information to return messages in the queried range. The details of the pagination are explained next.
+
+To perform pagination, each `WakuMessage` stored at a waku store node is associated with a unique `Index`.  Once a `WakuMessage` is received by a `WakuStore` node, it will generate the `Index` of the  `WakuMessage` and then saves the (`Index`, `WakuMessage`) together in its local storage. An `Index` encapsulates fields useful for paging purposes, these fields are a `digest` and a `receivedTime`. The `receivedTime` indicates the time at which the `WakuMessage` is received by the node and the `digest` is the hash of the received `WakuMessage`.  
+
+
+
+When a `HistoryQuery` arrives, it may contain an extra field `pagingInfo` that carries the information required for paging. The `pagingInfo` contains the `pageSize`, `cursor` and the `direction` as defined below:
+
+**pageSize**
+
+- A positive integer indicating the number of `WakuMessage`s to be included in the `HistoryResponse`
+- If the `pageSize` is omitted then a “default page size” shall be used
+- There is an upper limit for the page size
+- If the number of remaining `WakuMessage`s is less than the `pageSize`, then the `pageSize` is adjusted based on the remaining messages
+
+**Cursor**
+
+-  Holds the `Index` of a waku message
+
+**Direction**
+
+- Is a boolean value
+     - True means forward pagination that is the response contains `pageSize` many waku messages whose `Index` are larger than the given `cursor`
+     - False means backward pagination, that is the response contains `pageSize` many waku messages whose `Index` are smaller than or equal to the given cursor
+
+Given a valid  `pagingInfo`, the waku store node performs as follows:
+- sorts the list of waku messages based on their `Index`
+- finds the waku message whose index is the same as `cursor` 
+- slices `pageSize` many waku messages preceding or exceeding the `cursor` (depending on the `direction`)
+- the resultant list  is sent in the  `messages` field of the `HistoryResponse`
+
+The `HistoryResponse` will also contain an updated `pagingInfo` which helps the requester to resume its further requests; the `pageSize` indicates the number of returned waku messages, the `cursor` holds the `Index` of the last waku message in the returned list and the `direction` is the same direction as in the corresponding `HistoryQuery`.
+
+
+ <!---**Protocol identifier***: `/vac/waku/store/2.0.0-beta1`--->
+
+# Wire Specification
+
+Peers communicate with each other using a request/response API. The messages sent are Protobuf RPC messages. The followings are the Protobuf specifications. 
+
+## Protobuf
+
+```protobuf
+message Index {
```

Comment:
These would be part of the https://github.com/vacp2p/specs/blob/master/specs/waku/v2/waku-store.md#protobuf section

Reply:
Sure